editor/src/messages/tool/tool_messages/select_tool.rs

```diff
@@ -1590,16 +1645,47 @@ fn not_artboard(document: &DocumentMessageHandler) -> impl Fn(&LayerNodeIdentifi
 	|&layer| layer != LayerNodeIdentifier::ROOT_PARENT && !document.network_interface.is_artboard(&layer.to_node(), &[])
 }
 
-fn drag_shallowest_manipulation(responses: &mut VecDeque<Message>, selected: Vec<LayerNodeIdentifier>, tool_data: &mut SelectToolData, document: &DocumentMessageHandler) {
-	for layer in selected {
-		let ancestor = layer
-			.ancestors(document.metadata())
+fn drag_shallowest_manipulation(responses: &mut VecDeque<Message>, selected: Vec<LayerNodeIdentifier>, tool_data: &mut SelectToolData, document: &DocumentMessageHandler, remove: bool) {
+	if selected.is_empty() {
+		return;
+	}
+
+	let clicked_layer = document.find_deepest(&selected).unwrap_or_else(|| {
+		LayerNodeIdentifier::ROOT_PARENT
+			.children(document.metadata())
+			.next()
+			.expect("ROOT_PARENT should have at least one layer when clicking")
+	});
+
+	let metadata = document.metadata();
+
+	let selected_layers = document.network_interface.selected_nodes().selected_layers(document.metadata()).collect::<Vec<_>>();
+	let final_selection: Option<LayerNodeIdentifier> = (!selected_layers.is_empty() && selected_layers != vec![LayerNodeIdentifier::ROOT_PARENT]).then_some(()).and_then(|_| {
+		let mut relevant_layers = document.network_interface.selected_nodes().selected_layers(document.metadata()).collect::<Vec<_>>();
+		if !relevant_layers.contains(&clicked_layer) {
+			relevant_layers.push(clicked_layer);
+		}
+		clicked_layer
+			.ancestors(metadata)
 			.filter(not_artboard(document))
-			.find(|&ancestor| document.network_interface.selected_nodes().selected_layers_contains(ancestor, document.metadata()));
+			.find(|&potential_lca| relevant_layers.iter().all(|layer| *layer == potential_lca || potential_lca.is_ancestor_of(metadata, layer)))
+			.and_then(|lca| {
+				let lca_children: Vec<_> = lca.children(metadata).collect();
+				(clicked_layer == lca)
+					.then_some(lca)
+					.or_else(|| lca_children.iter().find(|&&child| clicked_layer == child || child.is_ancestor_of(metadata, &clicked_layer)).copied())
+			})
```

Comment:
What does `lca` stand for? Expand this please.

Reply:
Least common ancestor.